### ISSUE TITLE
Default initialization of variables

### ISSUE THREAD

**Issue by akimd** [koalaman/shellcheck#1084]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2086
- My shellcheck version (`shellcheck --version` or "online"): "online"
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#! /bin/bash

: ${BAR=a b c}
: ${FOO=$BAR}

echo "$FOO"
```

#### Here's what shellcheck currently says:
```
$ shellcheck /tmp/bar.sh

In /tmp/bar.sh line 3:
: ${BAR=a b c}
  ^-- SC2086: Double quote to prevent globbing and word splitting.


In /tmp/bar.sh line 4:
: ${FOO=$BAR}
  ^-- SC2086: Double quote to prevent globbing and word splitting.

```

#### Here's what I wanted or expected to see:

The use of `: ${v1=val} ${v2:=val}` to assign default values to variables is fairly common (e.g., the Autotools use this extensively), and I can't see how IFS splitting would matter in anyway here.  I agree this makes an exception and the less exceptions we are the happier we are, yet here, I think it would be worth it.

**Comment by nihilus** [comment]
Only because something is widely used doesnt mean i is good practice.

**Comment by nihilus** [comment]
Not to mention that autotools generate code using the m4 languge and that implicits using quirks to work around issues rather than generating readable scripts.

**Comment by akimd** [comment]
I agree that it's not because is widely used that it's necessarily good practice.  I disagree on the comment about M4: I was not only referring to the generated configure code (which does try to be readable), but also to the hand-written shell scripts that are libtool, autoconf, autoheader etc.

That being said, I would like to know what's wrong about this practice.

**Comment by Nightfirecat** [comment]
As far as I can tell, nothing's wrong with initializing variables a la `${v1=val}`, but the issue seems to be that it's reporting that `: ${BAR=a b c}` is equivalent to executing `: a b c`, that is `:` with three arguments, rather than the likely intended `: 'a b c'`. Given that `:` is a no-op command, it's not relevant in this example, but the warning is correct nonetheless: words would be split when being evaluated as-is.

**Comment by akimd** [comment]
@Nightfirecat I am precisely referring to this case: initializing variables using `:`.  In that particular case, which is a used idiom, I believe the warning is irrelevant, and "working around it" clutters the script.

**Comment by akimd** [comment]
Unless you meant that it's the same thing with `: $FOO`?  Well, in this case, the only sense I can make of it would be with `set -u`, and quotes, `: "$FOO"`, don't make it better (IMHO).

**Comment by kurahaupo** [comment]
It's not just word-splitting, it's globbing as well, which in some cases can be very expensive.

Consider:
```
shopt -s globstar
bar='/**/' # empty C comment
: ${foo:=$bar}
```

**Comment by akimd** [comment]
😄 😄 😄 

Got me.

**Comment by koalaman** [comment]
This comes up so often that it warrants a separate info message

**Comment by koalaman** [comment]
This will now produce a [SC2223](https://github.com/koalaman/shellcheck/wiki/SC2223) instead.

**Comment by akimd** [comment]
FWIW, I have been bitten by this.

I'm used to initialize my variable (after configuring) by sed-substitution this way:

```
: ${VCSN_PYTHONDIR='@PYEXECDIR@'}
```

and shellcheck pressed me to move to

```
: "${VCSN_PYTHONDIR='@PYEXECDIR@'}"
```

But then the meaning is completely different (I was amazed by that...): the inner single quotes now belong to the default value.

**Comment by koalaman** [comment]
Wow, I had no idea! The wiki now mentions this.

**Comment by kurahaupo** [comment]
I've taken to writing

    _=${foo=bar}
